### PR TITLE
BL-5122 prevent JS error by removing use of local storage

### PR DESCRIPTION
**Pull request by JohnThomson**
Still a kludge, because something goes wrong BEFORE we try to save data in local storage, but this may be a more robust way of preventing it from hurting us.

---
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/1903)

### CODE REVIEW

**Comment by hatton** [comment]
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/Edit/DecodableReaderTool.cs, line 79 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903#-KwlnOD7TAfWcBVKVN4p:-KwlnOD7TAfWcBVKVN4q:b-9ef4qz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/041f7dbb382643ad32b96f482c291588dd662c41/src/BloomExe/Edit/DecodableReaderTool.cs#L79)):*
> ```Smalltalk
> 		{
> 			base.SaveDefaultState();
> 			// We expect the state to be something like "stage:6;sort:byLength"
> ```

In other words, json without brackets? How about 

    let state = JSON.parse('{'+State+'}');
then, e.g.  ``state.stage`` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903)*

**Comment by JohnThomson** [comment]
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/Edit/DecodableReaderTool.cs, line 79 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903#-KwlnOD7TAfWcBVKVN4p:-Kwon8LQsy11jM1joT29:bgfd9mw) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/041f7dbb382643ad32b96f482c291588dd662c41/src/BloomExe/Edit/DecodableReaderTool.cs#L79)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

In other words, json without brackets? How about 

    let state = JSON.parse('{'+State+'}');
then, e.g.  ``state.stage`` 
</blockquote></details>

Unfortunately, it's not quite json...lacks quotes around values, and the separator is a semi-colon rather than a comma.
We could change it, but this syntax is in use in any decodable readers that have been already created, so we'd have to worry about data migration and loss of backwards compatibility...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903)*

**Comment by hatton** [comment]
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/Edit/DecodableReaderTool.cs, line 79 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903#-KwlnOD7TAfWcBVKVN4p:-KwoysCo4_PZ2WTbmoA_:btnxjqd) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/041f7dbb382643ad32b96f482c291588dd662c41/src/BloomExe/Edit/DecodableReaderTool.cs#L79)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Unfortunately, it's not quite json...lacks quotes around values, and the separator is a semi-colon rather than a comma.
We could change it, but this syntax is in use in any decodable readers that have been already created, so we'd have to worry about data migration and loss of backwards compatibility...
</blockquote></details>

Wow that is *insane*; why wasn't this done with json? We should indeed plan to fix that, even at the cost of backwards compatibility. That's just embarrassing.

I expect quotes won't matter here, there are no spaces in the data. So you could do

    let state = JSON.parse('{'+State.replace(';',',')+'}');

Or just use ``split()``?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903)*

**Comment by JohnThomson** [comment]
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/Edit/DecodableReaderTool.cs, line 79 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903#-KwlnOD7TAfWcBVKVN4p:-Kwp8fIM8pmovQ1Q36O2:b-o70vv0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/041f7dbb382643ad32b96f482c291588dd662c41/src/BloomExe/Edit/DecodableReaderTool.cs#L79)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Wow that is *insane*; why wasn't this done with json? We should indeed plan to fix that, even at the cost of backwards compatibility. That's just embarrassing.

I expect quotes won't matter here, there are no spaces in the data. So you could do

    let state = JSON.parse('{'+State.replace(';',',')+'}');

Or just use ``split()``?
</blockquote></details>

Possibly because this string gets embedded as a single value in a larger block of json, so using the json reserved characters within it would get a little messy? More likely it just seemed so simple it never occurred to me or whoever to introduce the complexity of a formal language? Anyway you're right, split is a much more concise way to express it so I've done that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1903)*

**Comment by hatton** [comment]
Since this is aimed at 4.0, it would require re-testing. I'm inclined to sit on it until we get 4.0 beta out the door.

**Comment by JohnThomson** [comment]
We wouldn't need to re-test EVERYTHING, I don't think...I'm quite sure it only affects Leveled/Decodable readers and 99% sure it only affects persisting the current level/stage. (But leaving till after the first beta is fine if you want.)
